core/services/relay/evm/batch_caller.go

```diff
@@ -0,0 +1,312 @@
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/client"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+)
+
+var errEmptyOutput = errors.New("rpc call output is empty (make sure that the contract method exists and rpc is healthy)")
+
+const (
+	// DefaultRpcBatchSizeLimit defines the maximum number of rpc requests to be included in a batch.
+	DefaultRpcBatchSizeLimit = 100
+
+	// DefaultRpcBatchBackOffMultiplier defines the rate of reducing the batch size limit for retried calls.
+	// For example if limit is 20 and multiplier is 4:
+	// 1.        20
+	// 2. 20/4 = 5
+	// 3. 5/4  = 1
+	DefaultRpcBatchBackOffMultiplier = 5
+
+	// DefaultMaxParallelRpcCalls defines the default maximum number of individual in-parallel rpc calls.
+	DefaultMaxParallelRpcCalls = 10
+)
+
+// BatchResult is organised by contracts names, key is contract name.
+type BatchResult map[string]ContractResults
+type ContractResults []MethodCallResult
+type MethodCallResult struct {
+	MethodName  string
+	ReturnValue any
+	Err         error
+}
+
+type BatchCall []Call
+type Call struct {
+	ContractAddress          common.Address
+	ContractName, MethodName string
+	Params, ReturnVal        any
+}
+
+func (c BatchCall) String() string {
+	callString := ""
+	for _, call := range c {
+		callString += fmt.Sprintf("%s\n", call.String())
+	}
+	return callString
+}
+
+// Implement the String method for the Call struct
+func (c Call) String() string {
+	return fmt.Sprintf("contractAddress: %s, contractName: %s, method: %s, params: %+v",
+		c.ContractAddress.Hex(), c.ContractName, c.MethodName, c.Params)
+}
+
+//go:generate mockery --quiet --name BatchCaller --output ./rpclibmocks --outpkg rpclibmocks --filename batch_caller.go --case=underscore
+type BatchCaller interface {
+	// BatchCall executes all the provided BatchRequest and returns the results in the same order
+	// of the calls. Pass blockNumber=0 to use the latest block.
+	BatchCall(ctx context.Context, blockNumber uint64, batchRequests BatchCall) (BatchResult, error)
+}
+
+// dynamicLimitedBatchCaller makes batched rpc calls and perform retries by reducing the batch size on each retry.
+type dynamicLimitedBatchCaller struct {
+	bc *defaultEvmBatchCaller
+}
+
+func NewDynamicLimitedBatchCaller(lggr logger.Logger, codec types.Codec, evmClient client.Client, batchSizeLimit, backOffMultiplier, parallelRpcCallsLimit uint) BatchCaller {
+	return &dynamicLimitedBatchCaller{
+		bc: newDefaultEvmBatchCaller(lggr, evmClient, codec, batchSizeLimit, backOffMultiplier, parallelRpcCallsLimit),
+	}
+}
+
+func (c *dynamicLimitedBatchCaller) BatchCall(ctx context.Context, blockNumber uint64, reqs BatchCall) (BatchResult, error) {
+	return c.bc.batchCallDynamicLimitRetries(ctx, blockNumber, reqs)
+}
+
+type defaultEvmBatchCaller struct {
```

Comment:
batchCaller and its tests are identical to what CCIP [is currently using](https://github.com/smartcontractkit/ccip/blob/54ee4f13143d3e414627b6a0b9f71d5dfade76c5/core/services/ocr2/plugins/ccip/internal/rpclib/evm.go#L75), with some adjustments to make it work better with Chain Reader interface

Reply:
This parallel batch caller is something now used by at least `ChainReader`, `CCIP`, `Automation`, and `Atlas` with various minor differences. We should make a common component to do this with various post-processing hooks for the batch results. Then we wouldn't have so much code duplication.

Reply:
Batch caller only exists in CCIP repo, can you link where Atlas and Automation use it? In any case, this implementation is similar enough to the batch caller that CCIP is using right now, so after next core release they can just migrate to this one

Reply:
I think that the work to do this is greater than maintaining duplicate code since batchCaller isn't something that is frequently updated